Sources/KIF/Additions/UIView-KIFAdditions.m

```diff
@@ -873,8 +873,8 @@ - (BOOL)isProbablyTappable
 // Is this view currently on screen?
 - (BOOL)isTappable;
 {
-    return ([self hasTapGestureRecognizerAndIsControlEnabled] ||
-            [self isTappableInRect:self.bounds]);
+    return ([self isTappableInRect:self.bounds] || ([self hasTapGestureRecognizerAndIsControlEnabled] &&
+            [self isTappableInRect:self.bounds]));
```

Comment:
Hello @justinseanmartin , I've found a bug here: a view was considered tappable even when occluded by another element, because `isTappable` was returning `true` even when only `hasTapGestureRecognizerAndIsControlEnabled` was true. 

Since this is a reliability issue, I thought to take advantage of the MR that has not been merged yet and fix it. I've added also integration tests to make sure it will be maintained over time 😀

I edited also the MR description, let me know if something is not clear 🙏🏻

Reply:
Was this breaking something for you, or just something you noticed that seemed wrong while you were in here?

I want to say that this was by design, because sometimes the tap event might trigger even if the element isn't the one directly that gets returned by a hit test. I'll verify by pointing our internal tests at this branch and see if anything breaks.

IIUC - I think the boolean logic you have here is equivalent to `return [self isTappableInRect:self.bounds];`, disregarding whether `hasTapGestureRecognizerAndIsControlEnabled` is true or false. If that's the case, we should simplify this, but I do think there are cases where this was necessary.

Reply:
Thanks for your answer! Indeed the boolean logic is the same, it can be simplified if we leave this change (sorry for that, just a bit tired 😅).
Yes it breaks `tappable` checks: calling `tryFindingTappableView()` I was getting `true` for occluded views. 
You can easily check it running the integration test I've added `testTryFindingOccludedTappableViewInFrameWithAccessibilityIdentifier` without this change. 

If you have a better idea on how to fix it, let me know and I can revert this change and follow your guidance 👍

Reply:
Looking at the history, we tried to remove this check in #1244 and ran into problems that led us to reverting that change in #1250.

I think the way I'd probably address this for now would be in `isTappable` to grab the intersection of the `self.bounds` rect with the view's window (might require translating coordinate systems) that pass that into `isTappableInRect:`. We could probably add a small optimization in `tappablePointInRect:` to early return with NO if the supplied rect has 0 by 0 dimensions.

Reply:
I confirmed that it does break a couple tests on a few screens in our app, but relatively minimal breakage. It might be something that we can work around on our end. In particular, I saw it trying to tap on a search field's placeholder text and saying it wasn't tappable. If we can avoid the breakage while still getting you the functionality that you need, that seems preferable.

Reply:
Ok, I'll try to follow your idea and commit the changes today, I'll keep you posted

Reply:
Sorry, I'm still digging into this to see if there is something that we can do on our end to work around it. This is in a custom search field control using [Blueprint](https://github.com/square/Blueprint). Given that it is only happening in one place in our codebase, it might be a problem with which view the accessibility identifier is being set on. Unfortunately I've got limited time today and am traveling the rest of the week, so it might be a little slow going on my end, but I will get back to looking at this next week.

If you can maintain the existing behavior regarding the gesture recognizers and only exclude offscreen elements in the tappability check, I think that'd be less likely to cause issues for existing tests. If not, no worries, just might take a little longer to dig through this on my end. In the meantime, you can depend on your unpublished KIF fork via git/branch to unblock yourself in the meantime and we'll get something that works here landed in the near future.

Reply:
Can we add a condition to check the [isVisibleInWindowFrame](https://github.com/kif-framework/KIF/blob/8757ba89034d7a590e383558908a44cc61c702bf/Sources/KIF/Additions/UIView-KIFAdditions.m#L1072) of this view if it isn't tappable but has a gesture recognizer?

Reply:
Hey Justin, thank you for your answers, sorry but I didn't have the time to check them before.

I've tried with a check like: 
```
([self isTappableInRect:self.bounds] || ([self hasTapGestureRecognizerAndIsControlEnabled] &&
            [self isVisibleInWindowFrame]));
```
but we have the same issue: An element that has a tap gesture recognizer and is visible in the current frame **but occluded by another view** is considered as tappable and this breaks some tests on our side.. 

So we still have the same issue, the element is actually not tappable (because totally occluded) but considered as it is 😢
I don't think there is a way to consider an element tappable when the check `isTappableInRect:self.bounds` fails, without affecting the reliability of this method tbh, but I might still miss something

What do you think?

Reply:
Ok, I figured out an issues with one of our test suites caused by this change.

We have a search bar element that contains a search text field as a subview. The text field is being marked with `userInteractionEnabled = false`, because the tap handler for the search bar redirects to another screen instead of allowing to type into it directly the text field. However, the accessibilityIdentifier we were matching on is being set on the search text field (matching what is shown in the text field placeholder). This works today because the element matching the hitTest (the search bar) has a tap gesture recognizer on it.

In our case, I can change the predicate to match on the accessibilityIdentifier of the search bar instead and things seem to work in this case. That said, this would also potentially break tests at other companies as well. I also still have another failure to look into still that isn't related to this case.

On the surface, it seems like it should be reasonable to allow the tap action to be handled by a superview that can process touch events. From a users perspective, they don't know or care if the tap is being handled by the search text field or the search bar. That said, trying to make `isTappableWithHitTestResultView:` allow for this seems to cause different problems, where we'll match on occluded elements that don't actually respond to touch events (like a label within a button).

I'm still thinking about this, but wanted to give a bit of an update here.

Reply:
Yeah, I got it, thank you for the update. We can keep each other updated here, I'll let you know if any idea comes to my mind when I'll have the time to work a bit on this

Reply:
Hey @justinseanmartin, I've experimented a bit more too but I wasn't able to find a solution to allow both cases 😞. 

_On a more personal note, I think it's reasonable in UITests to refer always to the actual element that should handle the tap (this is what we usually do, but as you mentioned I don't know if referring to a subview, which cannot handle tap events, is something usual in in other companies tests. The change might be justified in that case if we confirm there is no better solution to make the check reliable)._

Are you still trying to find a solution for this?

I'm sorry to push for an answer, the fact is that if we cannot merge the change for the reason above we will create a fork in our company GitHub account to avoid keeping reference to my personal account's fork
Thanks 🙏